### ISSUE TITLE
Expand problem for certain QG’s that result in disconnected Message KG’s

### ISSUE THREAD

**Issue by dkoslicki** [RTXteam/RTX#771]
It appears that `expand` is having issues with DSL queries such as:
```
add_qnode(curie=UniProtKB:P14136, id=n00)
add_qnode(curie=UniProtKB:P35579, id=n01)
add_qnode(type=biological_process, id=n02)
add_qedge(source_id=n00, target_id=n02, id=e00)
add_qedge(source_id=n01, target_id=n02, id=e01)
expand(edge_id=[e00,e01],kp=ARAX/KG1,continue_if_no_results=True)
resultify(ignore_edge_direction=True)
```
Which returns the result:
```
An uncaught error occurred: unhashable type: 'QEdge': ['Traceback (most recent call last):\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/swagger_server/../../../../ARAX/ARAXQuery/ARAX_query.py", line 491, in executeProcessingPlan\n result = expander.apply(message,action[\'parameters\'], response=response)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/swagger_server/../../../../ARAX/ARAXQuery/ARAX_expander.py", line 124, in apply\n answer_knowledge_graph = self.__expand_edge(edge_query_graph, kp_to_use)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/swagger_server/../../../../ARAX/ARAXQuery/ARAX_expander.py", line 247, in __expand_edge\n return kg_querier.answer_one_hop_query(query_graph)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/swagger_server/../../../../ARAX/ARAXQuery/Expand/kg_querier.py", line 56, in answer_one_hop_query\n self.__add_answers_to_kg(dsl_parameters[\'synonym_handling\'], synonym_usages_dict, kp, query_graph)\n', ' File "/mnt/data/orangeboard/production/RTX/code/UI/OpenAPI/python-flask-server/swagger_server/../../../../ARAX/ARAXQuery/Expand/kg_qu

```

Note that the equivalent cypher query on KG1: `with ["UniProtKB:P35579", "UniProtKB:P14136"] as inlist match p=(n:protein)-[]-(:biological_process) where n.id in inlist return p`
Returns a correct restful.

This was discovered by @chunyuma when addressing #623. Given it’s a synonym issue, @edeutsch may have something to add here...

**Comment by dkoslicki** [comment]
Note that “one at a time” works just fine:
```
add_qnode(curie=UniProtKB:P14136, id=n00)
add_qnode(type=biological_process, id=n02)
add_qedge(source_id=n00, target_id=n02, id=e00)
expand(edge_id=[e00],kp=ARAX/KG1)
resultify(ignore_edge_direction=True)
```
Returns without error, as well as:
```
add_qnode(curie=UniProtKB:P35579, id=n00)
add_qnode(type=biological_process, id=n02)
add_qedge(source_id=n00, target_id=n02, id=e00)
expand(edge_id=[e00],kp=ARAX/KG1)
resultify(ignore_edge_direction=True)
```

**Comment by dkoslicki** [comment]
Note @amykglen that @chunyuma is blocking on this for #623

**Comment by amykglen** [comment]
Got it - I believe I know what is throwing the error and will get it fixed. 

However, I do not think the 'equivalent cypher query' you listed is actually equivalent to this DSL. This is what the query graph you build in the DSL looks like:
```
(UniProtKB:P14136)--(biological_process)--(UniProtKB:P35579)
```
So (once the error is fixed) Expand will return all of the biological processes in common between those two proteins. But I believe your 'equivalent cypher query' returns all biological processes attached to _either_ of the two proteins. I don't think we currently have a way of representing such a query in a query graph. Perhaps we need to allow QNodes to have a list of curies? So that the DSL then looks like this:
```
add_qnode(curie=[UniProtKB:P14136, UniProtKB:P35579], id=n00)
add_qnode(type=biological_process, id=n01)
add_qedge(source_id=n00, target_id=n01, id=e00)
expand(kp=ARAX/KG1)
resultify()
```

**Comment by dkoslicki** [comment]
Ah ha, I see (and had overlooked that this DSL command made the biological process an intermediate node)! So @chunyuma should have one `biological_process` with a unique qnode id for each protein in this query and it should work as intended?

**Comment by amykglen** [comment]
True, that should work - but I think it doesn't yet, because in that case, some of the biological processes in the resulting knowledge_graph could have multiple qnode_ids... (whichever ones are in common between the two proteins) #720

The reported error is resolved in `master`, by the way. Though I think it (understandably?) errors out in `resultify` now, because `continue_if_no_results=true` was used, so the knowledge graph doesn't contain any complete paths matching the query graph (only partial).

**Comment by chunyuma** [comment]
Thanks @amykglen and @amykglen. @amykglen, you're right. I tried one `biological_process` with a unique qnode id for each protein but it seems like getting stuck in step 7. 

```
add_qnode(curie=UniProtKB:P14136, id=n00)
add_qnode(curie=UniProtKB:P35579, id=n01)
add_qnode(type=biological_process, id=n02)
add_qnode(type=biological_process, id=n03)
add_qedge(source_id=n00, target_id=n02, id=e00)
add_qedge(source_id=n01, target_id=n03, id=e01)
expand(edge_id=[e00,e01],kp=ARAX/KG1)
resultify(ignore_edge_direction=false)
```

**Comment by amykglen** [comment]
Ah, yeah - I guess in addition to the potential multiple qnode_id problem there, that's also a disconnected query graph, which I don't think `expand` is currently set up to handle. (That could be changed, of course... though I think I recall there being some debate over whether disconnected query graphs should be allowed? #640)

**Comment by edeutsch** [comment]
@amykglen, regarding your question about lists: YES, we do want to support lists at some point because this is highly relevant and useful.  But I think we need to work out how this should be expressed in the reasoner standard query_graph first. It's not obvious that that format supports this well yet. We could make curie a list. The whole is_set=true issue comes up again here too. Should there be different behavior if is_set=true or false if a qnode is a list?

Maybe this should be its own issue.

**Comment by edeutsch** [comment]
Regarding Chunya's most recent DSL and Amy's response and #640: Chunya's DSL is basically asking two separate questions simultaneously in parallel. Allowing this could open a whole huge can of worms. I suggest we return an error if any one node in an input query graph is not reachable to all others, as I recommended in #640, otherwise it's just not clear how to handle this. I'll note that there is not basic checking along these lines for incoming query_graphs, but it occurs to me that there is NO checking if a query_graph is valid if it is built with DSL. There probably should be some code along these lines, because people will write weird DSL that the rest of the code can't handle.

**Comment by edeutsch** [comment]
> @amykglen, regarding your question about lists: YES, we do want to support lists at some point because this is highly relevant and useful. But I think we need to work out how this should be expressed in the reasoner standard query_graph first. It's not obvious that that format supports this well yet. We could make curie a list. The whole is_set=true issue comes up again here too. Should there be different behavior if is_set=true or false if a qnode is a list?
> 
> Maybe this should be its own issue.

ah, it turns out this is already an issue. #717. It didn't gain much attention at the time..

**Comment by edeutsch** [comment]
and I am now fully remembering that #717 is high priority if you want to be able to ask questions that apply fisher exact test.

**Comment by dkoslicki** [comment]
@edeutsch Exactly: Previously @chunyuma has been making one DSL call per node in the KG of a given type to determine the count of it’s neighbors (needed for FET calculation in #623) which was much too slow when making >80 calls. So I thought of (before #717 is implemented), just making one massive QG that asks for all neighbors via the DSL in one go (resulting in a disconnected QG & KG, hence against the paradigm suggested by #772). Any idea of a way forward? Wait until #717 is implemented, or try to allow `expand` to disconnected QG/KG’s?

**Comment by edeutsch** [comment]
I cannot see any reasonable alternative to my proposal of #717. But I may be missing something.

My only idea of a way forward is for us as a group to decide that the #717 proposal is the correct design decision or decide on an alternative, and then we all implement it in our various components. I am puzzled why there is no discussion over at #717.

**Comment by amykglen** [comment]
@dkoslicki - I don't think the method of making one massive disconnected QG would actually be much faster than making >80 separate DSL calls, at least for Expand - because behind the scenes Expand would still have to make separate calls for each of those >80 edges, due to the structure of the query graph.

**Comment by chunyuma** [comment]
@amykglen, is it possible to use parallel running to call API for each of edges behind Expand so that it can make it faster?

**Comment by amykglen** [comment]
@chunyuma  - yes, we could for this disconnected query graph, but for most of the regular query graphs we see, we can't really expand edges in parallel, because usually the results from one edge are fed into the next edge.

so I'm not sure it would be worth pursuing that if this was just a temporary way of allowing fisher exact test to work - it seems it would make more sense to just address #717.

**Comment by chunyuma** [comment]
@amykglen I see. So @dkoslicki, it seems like calling DSL per node within FET in parallel is still the best way to speed up the running.

**Comment by amykglen** [comment]
I think we can close this issue now, since the original reported error is resolved, and #717 has been/is being addressed?

**Comment by dkoslicki** [comment]
@amykglen has this been tested in prod/beta/devED? I still see the following error: `ERROR:  the node binding list does not cover all nodes in the query graph` for the DSL:
```
add_qnode(curie=UniProtKB:P14136, id=n00)
add_qnode(curie=UniProtKB:P35579, id=n01)
add_qnode(type=biological_process, id=n02)
add_qedge(source_id=n00, target_id=n02, id=e00)
add_qedge(source_id=n01, target_id=n02, id=e01)
expand(edge_id=[e00,e01],kp=ARAX/KG1,continue_if_no_results=True)
resultify(ignore_edge_direction=True)
```

**Comment by amykglen** [comment]
Yes, that's an error `resultify` is throwing (which I mentioned [here](https://github.com/RTXteam/RTX/issues/771#issuecomment-633814687)) because the knowledge graph does not contain any complete paths (because `continue_if_no_results=True` was used) - how resultify should behave in such situations is being discussed in #774

**Comment by dkoslicki** [comment]
Ah, got it, thanks! Then yeah, it can be closed from my perspective
